README.md
# Virtua Docker Image - PHP

> [`virtuasa/php:5.2`](https://microbadger.com/images/virtuasa/php:5.2)
> [`virtuasa/php:5.3`](https://microbadger.com/images/virtuasa/php:5.3)
> [`virtuasa/php:5.4`](https://microbadger.com/images/virtuasa/php:5.4)
> [`virtuasa/php:5.5`](https://microbadger.com/images/virtuasa/php:5.5)
> [`virtuasa/php:5.6`](https://microbadger.com/images/virtuasa/php:5.6)
> [`virtuasa/php:7.0`](https://microbadger.com/images/virtuasa/php:7.0)
> [`virtuasa/php:7.1`](https://microbadger.com/images/virtuasa/php:7.1)
> [`virtuasa/php:7.2`](https://microbadger.com/images/virtuasa/php:7.2)

These images are built and published on [VirtuaSA Docker Hub](https://hub.docker.com/r/virtuasa/php/).

## Build

* Run the script `./docker-build.sh`:
  * `./docker-build.sh <PHP_VERSION>`
  * `./docker-build.sh all`
* Images can be pushed from GitlabCI if all tests are OK

## Images content

| [Tags (`virtuasa/php:`)](https://hub.docker.com/r/virtuasa/php/tags/)                             | 5.2      | 5.3       | 5.4      | 5.5      | 5.6      | 7.0       | 7.1       | 7.2       |
| ------------------------------------------------------------------------------------------------- | :------: | :-------: | :------: | :------: | :------: | :-------: | :-------: | :-------: |
| [PHP (`php`)](http://php.net/manual/en/)                                                          | *5.2.6*  | *5.3.3*   | *5.4.45* | *5.5.38* | *5.6.30* | *7.0.26*  | *7.1.13*  | *7.2.1*   |
| [Behat (`behat`)](http://behat.org/en/latest/quick_start.html)                                    |          |           |          | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [Composer (`composer`)](https://getcomposer.org/doc/01-basic-usage.md)                            |          | &#x2714;  | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [DbUnit (`dbunit`)](https://github.com/sebastianbergmann/dbunit)                                  |          |           |          |          |          | &#x2714;  | &#x2714;  | &#x2714;  |
| [Phing (`phing`)](https://www.phing.info/docs/guide/stable/)                                      |          | &#x2714;  | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [PHP_CodeSniffer (`phpcs`, `phpcbf`)](https://github.com/squizlabs/PHP_CodeSniffer/wiki/Usage)    |          |           | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [PHP Copy/Paste Detector (`phpcpd`)](https://github.com/sebastianbergmann/phpcpd)                 |          |           |          |          | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [PHP_Depend (`pdepend`)](https://pdepend.org/documentation/getting-started.html)                  |          | &#x2714;  | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [PhpDocumentor (`phpdoc`)](https://docs.phpdoc.org/)                                              |          |           | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [PHPLOC (`phploc`)](https://github.com/sebastianbergmann/phploc)                                  |          |           |          |          | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [PhpMetrics (`phpmetrics`)](http://www.phpmetrics.org/documentation/index.html)                   |          |           | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [PHP Mess Detector (`phpmd`)](https://phpmd.org/documentation/index.html)                         |          | &#x2714;  | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [PHPUnit (Current Stable Release) (`phpunit`)](https://phpunit.de/manual/current/en/index.html)   |          |           |          |          |          | &#x2714;  | &#x2714;  | &#x2714;  |
| [PHPUnit 6.2+ (`phpunit62`)](https://phpunit.de/manual/6.2/en/index.html)                         |          |           |          |          |          | &#x2714;  | &#x2714;  | &#x2714;  |
| [PHPUnit 5.7+ (`phpunit57`)](https://phpunit.de/manual/5.7/en/index.html)                         |          |           |          |          | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [PHPUnit 4.8+ (`phpunit48`)](https://phpunit.de/manual/4.8/en/index.html)                         |          | &#x2714;  | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [Node.js (`node`, `npm`)](https://nodejs.org/dist/latest-v8.x/docs/api/)                          |          |           | *6.9.1*  | *6.9.1*  | *6.9.1*  | *8.9.0*   | *8.9.0*   | *8.9.0*   |
| [Bower (`bower`)](https://bower.io/)                                                              |          |           | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [Gulp (`gulp`)](https://github.com/gulpjs/gulp/blob/master/docs/API.md)                           |          |           | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [Grunt (`grunt`)](https://gruntjs.com/getting-started)                                            |          |           | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [PM2 (`pm2`)](http://pm2.keymetrics.io/docs/usage/quick-start/)                                   |          |           | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [Webpack (`webpack`)](https://webpack.js.org/concepts/)                                           |          |           | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [Yarn (`yarn`)](https://yarnpkg.com/en/docs)                                                      |          |           | &#x2714; | &#x2714; | &#x2714; | &#x2714;  | &#x2714;  | &#x2714;  |
| [Apache](https://httpd.apache.org/docs/2.4/)                                                      | *2.2.9*  | *2.2.16*  | *2.2.22* | *2.2.22* | *2.4.10* | *2.4.25*  | *2.4.25*  | *2.4.25*  |
| [NGINX](https://nginx.org/en/docs/)                                                               | *0.6.32* | *0.7.67*  | *1.2.1*  | *1.2.1*  | *1.6.2*  | *1.10.3*  | *1.10.3*  | *1.10.3*  |
| [Debian](https://www.debian.org/doc/)                                                             | *lenny*  | *squeeze* | *wheezy* | *wheezy* | *jessie* | *stretch* | *stretch* | *stretch* |

To customize these setup on your projects, changes must only be done to file [`docker-compose.yml`](https://docs.docker.com/compose/compose-file/compose-file-v1/).

*Apache* calls PHP with *mod_php*. *Nginx* calls PHP with

## Usage

### Customizable environment variables

> These variable are used on runtime and can change the default behavior of the images.
> Their value can be changed in `docker-compose.yml` file.

| Variable name                   | Default value                             | Description                                                                    |
| ------------------------------- | ----------------------------------------- | ------------------------------------------------------------------------------ |
| `APACHE_DOCUMENT_ROOT`          | `web`                                     | Path to the Apache document root folder \*
| `APACHE_LOG_PATH`               | `var/logs/apache`                         | Path of the website Apache log folder \*
| `APACHE_RUN_GROUP`              | `docker`                                  | Group of user running Apache
| `APACHE_RUN_USER`               | `docker`                                  | User running Apache
| `APACHE_SERVER_ADMIN`           | `webmaster@localhost`                     | Apache server administrator email address
| `APACHE_SERVER_NAME`            | `localhost`                               | Apache server name
| `APACHE_SSL_CERT_KEY`           | `/etc/ssl/private/ssl-cert-snakeoil.key`  | Private key used by Apache for SSL \**
| `APACHE_SSL_CERT_PEM`           | `/etc/ssl/certs/ssl-cert-snakeoil.pem`    | Public key used by Apache for SSL \**
| `DOCKER_BASE_DIR`               | `/data`                                   | Docker mount path inside the container \**
| `DOCKER_CHMOD_666`              | ` ` *(empty)*                             | Execute a `chmod 666` on given files if set \*
| `DOCKER_CHMOD_777`              | ` ` *(empty)*                             | Execute a `chmod 777` on given files if set \*
| `DOCKER_CHMOD_R666`             | ` ` *(empty)*                             | Execute a `chmod -R 666` on given files if set \*
| `DOCKER_CHMOD_R777`             | ` ` *(empty)*                             | Execute a `chmod -R 777` on given files if set \*
| `DOCKER_COPY_CONFIG_FROM_HOST`  | `false`                                   | Copy the configuration files from the setup folder on the host if set to `true`
| `DOCKER_COPY_CONFIG_TO_HOST`    | `false`                                   | Copy the configuration files to the setup folder on the host if set to `true`
| `DOCKER_COPY_IP_TO_HOST`        | `false`                                   | Copy the container's ip to the setup folder on the host if set to `true`
| `DOCKER_CUSTOM_INIT`            | `docker-init.sh`                          | Execute script before doing anything else \*
| `DOCKER_CUSTOM_START`           | `docker-start.sh`                         | Execute script just before starting web server \*
| `DOCKER_CUSTOM_STOP`            | `docker-stop.sh`                          | Execute script before shutting down \*
| `DOCKER_DEBUG`                  | ` ` *(empty)*                             | Enable debug output if any value is set
| `DOCKER_HOST_GID`               | ` ` *(empty)*                             | `chown` the mount path to given GID (see `id -g`) if set \***
| `DOCKER_HOST_SETUP_DIR`         | `setup`                                   | Path of the setup configuration files on the host \*
| `DOCKER_HOST_UID`               | ` ` *(empty)*                             | `chown` the mount path to given UID (see `id -u`) if set \***
| `DOCKER_MKDIR`                  | ` ` *(empty)*                             | Create requested directories using `mkdir -p` if set \*
| `DOCKER_TIMEZONE`               | `Europe/Zurich`                           | Time zone of the Docker container
| `DOCKER_WEB_SERVER`             | `apache`                                  | Web server to use, can be either `apache` or `nginx`
| `NGINX_DOCUMENT_ROOT`           | `web`                                     | Path to the Nginx document root folder \*
| `NGINX_LOG_PATH`                | `var/logs/nginx`                          | Path of the website Nginx log folder \*
| `NGINX_RUN_GROUP`               | `docker`                                  | Group of user running Nginx
| `NGINX_RUN_USER`                | `docker`                                  | User running Nginx
| `PHP_LOG_PATH`                  | `var/logs/php`                            | Path of the PHP log folder \*
| `PHP_MEMORY_LIMIT_APACHE`       | `128M`                                    | Maximum amount of memory a PHP script running on Apache may consume
| `PHP_MEMORY_LIMIT_CLI`          | `512M`                                    | Maximum amount of memory a PHP script running on CLI may consume
| `PHP_XDEBUG_APACHE_ENABLE`      | `off`                                     | Force activation of PHP XDebug on Apache if set to `on`
| `PHP_XDEBUG_CLI_ENABLE`         | `off`                                     | Force activation of PHP XDebug on CLI if set to `on`

\* *Path is relative to the Docker container mount path, i.e., to the root of the project.*

\*\* *Path is absolute in the Docker container, don't forget to add the path of the mount path if needed.*

\*\*\* *`DOCKER_HOST_GID` and `DOCKER_HOST_UID` must be both empty or both have a value.*

### Readonly environment variables

> These variables are used to build the images and can be used to adapt behaviors between their different flavors.
> *Changing their value is not recommanded.*

| Variable name                   | Sample value                              | Description                                                                    |
| ------------------------------- | ----------------------------------------- | ------------------------------------------------------------------------------ |
| `DOCKER_BUILD_DATE`             | `2017-12-13T15:22:29+0100`                | Build date of the image, using ISO-8601 format
| `DOCKER_FROM_COMMIT`            | `f84a27d7`                                | Short hash of the git commit used to build the Docker image in this repository
| `DOCKER_FROM_IMAGE`             | `debian/stretch`                          | Name of the base image used to build the Docker image
| `PHP_VERSION`                   | `7.1`                                     | PHP version included in the image
| `PHP_VERSION_APT`               | `7.1`                                     | PHP version used in the `apt-get install` instructions
| `PHP_VERSION_DIR`               | `/7.1`                                    | PHP version used in the paths of PHP configuration files

### Limitations with old PHP versions

* `PHP 5.2`:
  * Using `DOCKER_COPY_CONFIG_TO_HOST=true` will override host local files sharring the same names on each startup of the container.
  * Using `DOCKER_WEB_SERVER=nginx` is not allowed, and `apache` will be used instead, this can be overriden by setting the value of `DOCKER_WEB_SERVER` with a script loaded by `DOCKER_CUSTOM_INIT`.
* `PHP 5.3`:
  * Using `DOCKER_WEB_SERVER=nginx` is not allowed, and `apache` will be used instead, this can be overriden by setting the value of `DOCKER_WEB_SERVER` with a script loaded by `DOCKER_CUSTOM_INIT`.

## Changelog

See [CHANGELOG.md](CHANGELOG.md).

CHANGELOG.md
# Virtua Docker Container - PHP - CHANGELOG

## V1.3.8 - 20180108

* Metadata
  * Version: 1.3.8
  * Tags: `virtuasa/php:*-20180108`
  * Built on commit : [008d6b18](https://gitlab.virtua.ch/docker/virtuasa-php/tree/008d6b1871f34ab120b77a7ced7c510e6649ca68)
* Changes
  * Added: PHP versions upgraded: 7.0.26, 7.1.13, 7.2.1

## V1.3.7 - 20171213

* Metadata
  * Version: 1.3.7
  * Tags: `virtuasa/php:*-20171213`
  * Built on commit : [3b0cf86d](https://gitlab.virtua.ch/docker/virtuasa-php/tree/3b0cf86d818126fc02e6223a4058cb6d75455c44)
* Changes
  * Added: `DOCKER_MKDIR` to recursively create folders on start
  * Added: Labels compliant to http://label-schema.org/rc1/

## V1.3.6 - 20171208

* Metadata
  * Version: 1.3.6
  * Tags: `virtuasa/php:*-20171208`
  * Built on commit : [31849eea](https://gitlab.virtua.ch/docker/virtuasa-php/tree/31849eea9a7a99633c3e8bd81a27e35d51a99d27)
* Changes
  * Added: Ruby and Capistrano for PHP 5.6, 7.0, 7.1, 7.2

## V1.3.5 - 20171204

* Metadata
  * Version: 1.3.5
  * Tags: `virtuasa/php:*-20171204`
  * Built on commit : [0a1209e2](https://gitlab.virtua.ch/docker/virtuasa-php/tree/0a1209e2890aa0229821babb381804d1ade98b83)
* Changes
  * Added: Nginx is now working and can be selected with `DOCKER_WEB_SERVER=nginx` on PHP 7.0, 7.1, 7.2
  * Added: `sendmail` package
  * Updated: PHP 7.0 is now final

## V1.3.4 - 20171130

* Metadata
  * Version: 1.3.4
  * Tags: `virtuasa/php:*-20171130`
  * Built on commit : [1cf333d9](https://gitlab.virtua.ch/docker/virtuasa-php/tree/1cf333d97a90cb9b25d939e6527be2f584dac89c)
* Changes
  * Changed: Use real structure of tool configuration files instead of the old flat one
  * Fixed: Split PHP configuration files between apache, cli, fpm

## V1.3.3 - 20171129

* Metadata
  * Version: 1.3.3
  * Tags: `virtuasa/php:*-20171129`
  * Built on commit : [d5475ca0](https://gitlab.virtua.ch/docker/virtuasa-php/tree/d5475ca00f9d7423c9373629b3e3f4ba52c7adcf)
* Changes
  * Improved: More tests on images
  * Fixed: Images can work with and without debug mode
  * Fixed: Copy config to host works on PHP 5.2, but it overrides existing files

## V1.3.2 - 20171128

* Metadata
  * Version: 1.3.2
  * Tags: `virtuasa/php:*-20171128`
  * Built on commit : [b4bd8651](https://gitlab.virtua.ch/docker/virtuasa-php/tree/b4bd86513ef741a1df537cecb7b2d654e6160fb5)
* Changes
  * Fixed: `PHP_MEMORY_LIMIT_APACHE`, `PHP_MEMORY_LIMIT_CLI` are now always used by PHP
  * Fixed: No more issue when copying configuration files to / from host
  * Improved: Web server check to use is done before calling custom init script, so it can be overriden
  * Improved: Use templates for PHP configuration files too

## V1.3.1 - 20171126

* Metadata
  * Version: 1.3.1
  * Tags: `virtuasa/php:*-20171126`
  * Built on commit : [fcf7fc3b](https://gitlab.virtua.ch/docker/virtuasa-php/tree/fcf7fc3b45bdeabe3b217013128f52eb1becb87c)
* Changes
  * Added: `PHP_MEMORY_LIMIT_APACHE`, `PHP_MEMORY_LIMIT_CLI` to set maximum memory a PHP script may consumes
  * Removed: `APACHE_LOG_DIR` environment variable (use same path as `APACHE_LOG_PATH`)

## v1.3.0 - 20171123

* Metadata
  * Version: 1.3.0
  * Tags: `virtuasa/php:*-20171123`
  * Built on commit : [03de4045](https://gitlab.virtua.ch/docker/virtuasa-php/tree/03de4045f49660e279bed14a26b42ed946767e26)
* Changes
  * Fixed: no more issues when exporting environment variables to Apache
  * Added: ability to use Nginx instead of Apache (experimental)
  * Added: `envsubst`, `unzip` packages
  * Added: `DOCKER_CUSTOM_STOP` to run script on shutting down
  * Improved: declare volume in `Dockerfile`
  * Improved: `DOCKER_BASE_DIR` now changes working directory too
  * Improved: file permissions are now reset on shutting down too
  * Removed: `phing`, `pdepends` for PHP 5.2
  * Removed: `behat`, `phpdoc` for PHP 5.3
  * Removed: `behat` for PHP 5.4

## v1.2.0 - 20171103

* Metadata
  * Version: 1.2.0
  * Tags: `virtuasa/php:*-20171103`
  * Built on commit : [53312599](https://gitlab.virtua.ch/docker/virtuasa-php/tree/53312599c2af78cacedff62564ca58ea261b2fd6)
* Changes
  * Changed: use Node.js upgraded from 6.9.1 to 8.9.0 LTS
  * Added: `behat`, `dbunit`, `phing`, `phpdoc`, `phploc`, `phpmetrics`

## v1.1.3 - 20171102

* Metadata
  * Version: 1.1.3
  * Tags: `virtuasa/php:*-20171102`
  * Built on commit : [2976a595](https://gitlab.virtua.ch/docker/virtuasa-php/tree/2976a59513035a3dcdd69f8afa531d692800a5cc)
* Changes
  * Fixed: no more warning on PHP 5.6 about `php5-uprofiler` (`profiler.so` missing)
  * Added: `pdepend`, `phpcbf`, `phpcpd`, `phpcs`, `phpmd`, `pm2`
  * Added: short hashcode of git commit used to build images is displayed on startup

## v1.1.2 - 20171026

* Metadata
  * Version: 1.1.2
  * Tags: `virtuasa/php:*-20171026`
  * Built on commit : [3a20b13c](https://gitlab.virtua.ch/docker/virtuasa-php/tree/3a20b13cb4e128189e8e73e6f01bfc707d210de9)
* Changes
  * Added: `DOCKER_CUSTOM_INIT` environment variable to execute a script at initialization
  * Added: `DOCKER_CHMOD_R666` environment variable to `chmod -R 666` specified files
  * Added: `DOCKER_CHMOD_R777` environment variable to `chmod -R 777` specified files
  * Added: `DOCKER_CHMOD_666` environment variable to `chmod 666` specified files
  * Added: `DOCKER_CHMOD_777` environment variable to `chmod 777` specified files

## v1.1.1 - 20171025

* Metadata
  * Version: 1.1.1
  * Tags: `virtuasa/php:*-20171025`
  * Built on commit : [d625fa8d](https://gitlab.virtua.ch/docker/virtuasa-php/tree/d625fa8d3e671dc8a0dc83f26cbd238d50f2f05b)
* Changes
  * Changed: installation files moved to `/setup`

## v1.1.0 - 20171023

* Metadata
  * Version: 1.1.0
  * Tags: `virtuasa/php:*-20171023`
  * Built on commit : [0f5dbd79](https://gitlab.virtua.ch/docker/virtuasa-php/tree/0f5dbd79051c4f7af3c990c91f947c2fe53b6e38)
* Changes
  * Fixed: php5-snmp now works well on PHP 5.6
  * Fixed: php files are now correctly handled by Apache with PHP 5.5

setup/docker/start.sh
#!/bin/bash
set -e

echo "Starting the Virtua Docker Container ..."
echo "More info at: <https://hub.docker.com/r/virtuasa/php/>"
echo "Built from commit: ${DOCKER_FROM_COMMIT}"
echo

# Print ran commands
[[ -n "${DOCKER_DEBUG}" ]] && set -x

# Set working directory
sudo mkdir -p ${DOCKER_BASE_DIR}
cd ${DOCKER_BASE_DIR}

# Disable Nginx for PHP < 7.0
[[ "${PHP_VERSION}" =~ ^(5\.[23456]) ]] && export DOCKER_WEB_SERVER="apache"

# Exec custom init script
[[ -n "${DOCKER_CUSTOM_INIT}" ]] && [[ -e "${DOCKER_CUSTOM_INIT}" ]] && sudo chmod +x "${DOCKER_CUSTOM_INIT}" \
    && echo "Executing custom init script: ${DOCKER_CUSTOM_INIT} ..." \
    && ./${DOCKER_CUSTOM_INIT}

# Set trap for every signals
trap 'echo -e "\n(last error code: '"'"'$?/${PIPESTATUS[*]}'"'"'; on line: '"'"'${LINENO}/$(caller)'"'"'; with command: '"'"'${BASH_COMMAND}'"'"')";
    PTK="${!}"; [[ -n "${PTK}" ]] && ps -e | grep -q "${PTK}" && kill ${PTK};
    echo "Stopping the Virtua Docker Container ..."; . /setup/docker/stop.sh' 0

# Set timezone
echo "${DOCKER_TIMEZONE}" | sudo tee /etc/timezone > /dev/null
sudo dpkg-reconfigure --frontend noninteractive tzdata

# Print installed PHP version
php -v

# Get HOST ip address
DOCKER_HOST_IP="${DOCKER_HOST_IP:-$(/sbin/ip route | awk '/default/ { print $3 }')}"
export DOCKER_HOST_IP

# Print all environment variables (can be overriden in docker-compose.yml)
[[ -n "${DOCKER_DEBUG}" ]] && printenv

# Print all shell options
[[ -n "${DOCKER_DEBUG}" ]] && shopt

# Enable some shell options
shopt -s extglob

# Force the UID/GID of the docker container running user
[[ -n "${DOCKER_USER_UID}" ]] && usermod -u ${DOCKER_USER_UID} docker
[[ -n "${DOCKER_USER_GID}" ]] && groupmod -u ${DOCKER_USER_GID} docker && usermod -g ${DOCKER_USER_GID} docker

# Copy Apache configuration files
sudo rm -rf /etc/apache2/sites-available/*
sudo cp -r /setup/apache/* /etc/apache2

# Copy Nginx configuration files
sudo rm -rf /etc/nginx/nginx.conf*
sudo rm -rf /etc/nginx/sites-available/*
sudo rm -rf /etc/nginx/sites-enabled/*
sudo cp -r /setup/nginx/* /etc/nginx

# Copy PHP configuration files
sudo cp -r /setup/php/apache/* /etc/php${PHP_VERSION_DIR}/apache2
sudo cp -r /setup/php/cli/* /etc/php${PHP_VERSION_DIR}/cli
sudo cp -r /setup/php/fpm/* /etc/php${PHP_VERSION_DIR}/fpm


# Copy image's configuration files to host filesystem
if [[ "${DOCKER_COPY_CONFIG_TO_HOST}" = "true" ]]; then
    sudo mkdir -p "${DOCKER_HOST_SETUP_DIR}/apache"
    sudo mkdir -p "${DOCKER_HOST_SETUP_DIR}/docker"
    sudo mkdir -p "${DOCKER_HOST_SETUP_DIR}/nginx"
    sudo mkdir -p "${DOCKER_HOST_SETUP_DIR}/php/apache"
    sudo mkdir -p "${DOCKER_HOST_SETUP_DIR}/php/cli"
    sudo mkdir -p "${DOCKER_HOST_SETUP_DIR}/php/fpm"
    if [[ "${DOCKER_FROM_IMAGE##*:}" = "lenny" ]]; then
        sudo cp -r "/etc/apache2/"* "${DOCKER_HOST_SETUP_DIR}/apache"
        sudo cp -r "/setup/docker/.gitignore" "${DOCKER_HOST_SETUP_DIR}/docker"
        sudo cp -r "/etc/nginx/"* "${DOCKER_HOST_SETUP_DIR}/nginx"
        sudo cp -r "/etc/php${PHP_VERSION_DIR}/apache2/"* "${DOCKER_HOST_SETUP_DIR}/php/apache"
        sudo cp -r "/etc/php${PHP_VERSION_DIR}/cli/"* "${DOCKER_HOST_SETUP_DIR}/php/cli"
        sudo cp -r "/etc/php${PHP_VERSION_DIR}/fpm/"* "${DOCKER_HOST_SETUP_DIR}/php/fpm"
    else
        sudo cp -nr "/etc/apache2/"* "${DOCKER_HOST_SETUP_DIR}/apache"
        sudo cp -nr "/setup/docker/.gitignore" "${DOCKER_HOST_SETUP_DIR}/docker"
        sudo cp -nr "/etc/nginx/"* "${DOCKER_HOST_SETUP_DIR}/nginx"
        sudo cp -nr "/etc/php${PHP_VERSION_DIR}/apache2/"* "${DOCKER_HOST_SETUP_DIR}/php/apache"
        sudo cp -nr "/etc/php${PHP_VERSION_DIR}/cli/"* "${DOCKER_HOST_SETUP_DIR}/php/cli"
        sudo cp -nr "/etc/php${PHP_VERSION_DIR}/fpm/"* "${DOCKER_HOST_SETUP_DIR}/php/fpm"
    fi
fi

# Copy image's configuration files from host filesystem
if [[ "${DOCKER_COPY_CONFIG_FROM_HOST}" = "true" ]]; then
    sudo cp -rf "${DOCKER_HOST_SETUP_DIR}/apache/"* "/etc/apache2"
    sudo cp -rf "${DOCKER_HOST_SETUP_DIR}/nginx/"* "/etc/nginx"
    sudo cp -rf "${DOCKER_HOST_SETUP_DIR}/php/apache/"* "/etc/php${PHP_VERSION_DIR}/apache2"
    sudo cp -rf "${DOCKER_HOST_SETUP_DIR}/php/cli/"* "/etc/php${PHP_VERSION_DIR}/cli"
    sudo cp -rf "${DOCKER_HOST_SETUP_DIR}/php/fpm/"* "/etc/php${PHP_VERSION_DIR}/fpm"
fi

# Chown the mount directory
[[ -n "${DOCKER_HOST_UID}" ]] && [[ -n "${DOCKER_HOST_GID}" ]] && sudo chown -R ${DOCKER_HOST_UID}:${DOCKER_HOST_GID} "${DOCKER_BASE_DIR}"

# Create the requested directories
[[ -n "${DOCKER_MKDIR}" ]] && sudo mkdir -p ${DOCKER_MKDIR}

# Chmod the requested directories
[[ -n "${DOCKER_CHMOD_666}" ]] && sudo chmod 666 ${DOCKER_CHMOD_666}
[[ -n "${DOCKER_CHMOD_777}" ]] && sudo chmod 777 ${DOCKER_CHMOD_777}
[[ -n "${DOCKER_CHMOD_R666}" ]] && sudo chmod -R 666 ${DOCKER_CHMOD_R666}
[[ -n "${DOCKER_CHMOD_R777}" ]] && sudo chmod -R 777 ${DOCKER_CHMOD_R777}

# Run Composer if necessary
if [[ "${PHP_VERSION}" != "5.2" ]]; then
    [[ -e 'composer.json' && ! -e 'vendor/autoload.php' ]] && composer install --no-interaction
fi

# Install npm packages if necessary
[[ -e 'packages.json' && ! -f 'node_modules' ]] && npm install

# Configure web server
if [[ "${DOCKER_WEB_SERVER}" = "apache" ]]; then
    # Apache gets grumpy about PID files pre-existing
    sudo rm -f /var/run/apache2/apache2.pid
    sudo rm -f /var/run/apache2/ssl_mutex
    sudo mkdir -p /var/run/apache2
    [[ "${PHP_VERSION}" = "5.4" ]] && sudo chown ${APACHE_RUN_USER} /var/lock/apache2
    # Apache log directory
    export APACHE_LOG_DIR="${DOCKER_BASE_DIR}/${APACHE_LOG_PATH}"
    sudo mkdir -p "${DOCKER_BASE_DIR}/${APACHE_LOG_PATH}"
    sudo chmod -R 755 "${DOCKER_BASE_DIR}/${APACHE_LOG_PATH}"
    # Apache root directory
    sudo mkdir -p "${DOCKER_BASE_DIR}/${APACHE_DOCUMENT_ROOT}"
    export DOLLAR='$'
    # Disable previous Apache sites
    (cd /etc/apache2/sites-enabled && find -mindepth 1 -print -quit | grep -q . && sudo a2dissite * || true)
    # Replace system environment variables into Apache configuration files
    echo -n "Applying Apache configuration file templates "
    find /etc/apache2 -name "*.conf.tpl" | while IFS= read -r file; do
        envsubst < ${file} | sudo tee ${file%%.tpl} > /dev/null
        sudo rm ${file}
        [[ -n "${DOCKER_DEBUG}" ]] && cat ${file%%.tpl}
        echo -n "."
    done
    echo " OK"
    # Configure Apache
    (cd /etc/apache2/sites-enabled && sudo a2ensite *)
    echo -e "\nexport APACHE_RUN_USER='${APACHE_RUN_USER}'\n" | sudo tee -a /etc/apache2/envvars > /dev/null
    echo -e "\nexport APACHE_RUN_GROUP='${APACHE_RUN_GROUP}'\n" | sudo tee -a /etc/apache2/envvars > /dev/null
    echo -e "\nexport APACHE_LOG_DIR='${APACHE_LOG_DIR}'\n" | sudo tee -a /etc/apache2/envvars > /dev/null
elif [[ "${DOCKER_WEB_SERVER}" = "nginx" ]]; then
    # Nginx log directory
    sudo mkdir -p "${DOCKER_BASE_DIR}/${NGINX_LOG_PATH}"
    sudo chmod -R 755 "${DOCKER_BASE_DIR}/${NGINX_LOG_PATH}"
    # Replace system environment variables into Nginx configuration files
    echo -n "Applying Nginx configuration file templates "
    find /etc/nginx -name "*.conf.tpl" | while IFS= read -r file; do
        envsubst < ${file} | sudo tee ${file%%.tpl} > /dev/null
        sudo rm ${file}
        [[ -n "${DOCKER_DEBUG}" ]] && cat ${file%%.tpl}
        echo -n "."
    done
    echo " OK"
fi

# PHP log direcotry
sudo mkdir -p "${DOCKER_BASE_DIR}/${PHP_LOG_PATH}"
sudo chmod -R 755 "${DOCKER_BASE_DIR}/${PHP_LOG_PATH}"

# Replace system environment variables into PHP configuration files
echo -n "Applying PHP configuration file templates "
find /etc/php${PHP_VERSION_DIR} -regex ".*\.\(conf\|ini\)\.tpl" | while IFS= read -r file; do
    envsubst < ${file} | sudo tee ${file%%.tpl} > /dev/null
    sudo rm ${file}
    [[ -n "${DOCKER_DEBUG}" ]] && cat ${file%%.tpl}
    echo -n "."
done
echo " OK"

# Start PHP-FPM if using Nginx
[[ "${DOCKER_WEB_SERVER}" = "nginx" ]] && sudo service php${PHP_VERSION_APT}-fpm start

# Clean Behat cache directory
sudo rm -rf /tmp/behat_gherkin_cache

# Exec custom startup script
[[ -n "${DOCKER_CUSTOM_START}" ]] && [[ -e "${DOCKER_CUSTOM_START}" ]] && sudo chmod +x "${DOCKER_CUSTOM_START}" \
    && echo "Executing custom start script: ${DOCKER_CUSTOM_START} ..." \
    && ./${DOCKER_CUSTOM_START}

# Display server IP
echo "Started ${DOCKER_WEB_SERVER} web server on ..."
[[ "${DOCKER_FROM_IMAGE##*:}" = "lenny" ]] \
    && IP="$(ifconfig | awk '/inet addr/{print substr($2,6)}' | head -n 3 | tail -n 1)" \
    || IP="$(hostname -I | cut -d' ' -f1)"
echo "> http://${IP}"
echo "> https://${IP}"
if [[ "${DOCKER_COPY_IP_TO_HOST}" = "true" ]]; then
    sudo mkdir -p "${DOCKER_HOST_SETUP_DIR}/docker"
    sudo cp -nr "/setup/docker/.gitignore" "${DOCKER_HOST_SETUP_DIR}/docker"
    echo "${IP}" | sudo tee "${DOCKER_HOST_SETUP_DIR}/docker/ip" > /dev/null
fi

# Chown the mount dir after Apache has started
(sleep 5s; [[ -n "${DOCKER_HOST_UID}" ]] && [[ -n "${DOCKER_HOST_GID}" ]] && sudo chown -R ${DOCKER_HOST_UID}:${DOCKER_HOST_GID} "${DOCKER_BASE_DIR}") &

# Start web server
if [[ "${DOCKER_WEB_SERVER}" = "apache" ]]; then
    [[ "${DOCKER_FROM_IMAGE##*:}" = "lenny" ]] \
        && sudo /usr/sbin/apache2ctl -DFOREGROUND \
        || sudo /usr/sbin/apache2ctl -D FOREGROUND
elif [[ "${DOCKER_WEB_SERVER}" = "nginx" ]]; then
    sudo nginx
fi